xds/internal/balancer/cdsbalancer/cdsbalancer.go

```diff
@@ -365,3 +504,39 @@ func (b *cdsBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.Sub
 func (b *cdsBalancer) Close() {
 	b.closed.Fire()
 }
+
+// ccWrapper wraps the balancer.ClientConn that was passed in to the CDS
+// balancer during creation and intercepts the NewSubConn() call from the child
+// policy. Other methods of the balancer.ClientConn interface are not overridden
+// and hence get the original implementation.
+type ccWrapper struct {
+	balancer.ClientConn
+
+	mu    sync.Mutex
+	xdsHI *xds.HandshakeInfo
+}
+
+// NewSubConn intercepts NewSubConn() calls from the child policy and adds an
+// address attribute which provides all information required by the xdsCreds
+// handshaker to perform the TLS handshake.
+func (ccw *ccWrapper) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	ccw.mu.Lock()
+	defer ccw.mu.Unlock()
+
+	if ccw.xdsHI == nil {
+		return ccw.ClientConn.NewSubConn(addrs, opts)
+	}
+
+	newAddrs := make([]resolver.Address, len(addrs))
+	for i, addr := range addrs {
+		newAddrs[i] = xds.SetHandshakeInfo(addr, ccw.xdsHI)
+	}
+	return ccw.ClientConn.NewSubConn(newAddrs, opts)
+}
+
+// setHandshakeInfo updates the handshakeInfo within the wrapper.
+func (ccw *ccWrapper) setHandshakeInfo(hi *xds.HandshakeInfo) {
+	ccw.mu.Lock()
```

Comment:
Is it possible to eliminate the lock by recreating the `ccWrapper` when the HI changes?  I.e. `ccWrapper` is immutable.

Reply:
We pass this `ccWrapper` as the `balancer.ClientConn` to the child policy. So, we could have a case where only the security config in the received `Cluster` resource changes (and therefore only the `HandshakeInfo` changes). There is no way to ask the child policy to use a new `balancer.ClientConn` in this case.

Reply:
Oh I see. I thought there was another level of indirection here:

child -> childCCWrapper -> CDS -> this `ccWrapper` -> parent

That's fine, though.  I think we should still be able to use an `atomic.Value` for `xdsHI` and avoid the mutex anyway, right?  Just make sure to `Load` it at the top everything is consistent.

Reply:
We would not need this mutex if we make the other change that I suggested (`ccWrapper` always configured with a `xds.HandshakeInfo` and we only change the fields inside the latter.